File: src/main/java/personal/alexgarland/helloworldmvc/service/ArrayListEmployeeRepository.java
package personal.alexgarland.helloworldmvc.service;

import static java.util.stream.Collectors.toList;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import personal.alexgarland.helloworldmvc.model.Employee;

public final class ArrayListEmployeeRepository implements IEmployeeRepository {

  private static List<Employee> employeeList;

  private static final IdSequence sequence;

  static {
    employeeList = new ArrayList<Employee>();
    employeeList.add(new Employee(1, "Michael", "Smith", "Mikey"));
    employeeList.add(new Employee(2, "Jonathan", "Taylor", "Johnny"));
    employeeList.add(new Employee(3, "David", "Wilson", "Dave"));
    sequence = new IdSequence(4);
  }

  @Override
  public List<Employee> getEmployeeList() {
    return employeeList;
  }

  @Override
  public Employee addEmployee(Employee e) {
    Employee employee = e.copyWithNewId(sequence.getNextId());
    employeeList.add(employee);
    return employee;
  }

  private List<Employee> listExcludingEmployee(int employeeId) {
    return employeeList.stream().filter(e -> e.getId() != employeeId).collect(toList());
  }

  @Override
  public void deleteEmployeeById(int employeeId) {
    employeeList = listExcludingEmployee(employeeId);
  }

  @Override
  public void updateEmployee(Employee e) {
    List<Employee> newList = listExcludingEmployee(e.getId());
    newList.add(e);
    Collections.sort(newList);
    employeeList = newList;
  }

  @Override
  public Employee getEmployeeById(int employeeId) {
    return employeeList.stream().filter(e -> e.getId() == employeeId).findFirst().get();
  }

}

File: src/main/kotlin/personal/alexgarland/helloworldmvc/service/ArrayListEmployeeRepository.kt
package personal.alexgarland.helloworldmvc.service

import personal.alexgarland.helloworldmvc.model.Employee

class ArrayListEmployeeRepository : IEmployeeRepository {

    override val employeeList: List<Employee>
        get() {
            return staticEmployeeList
        }

    override fun addEmployee(e: Employee): Employee {
        val employee = e.copy(id = sequence.getNextId())
        staticEmployeeList.add(employee)
        return employee
    }

    override fun deleteEmployeeById(employeeId: Int) {
        staticEmployeeList.removeAll({ it.id != employeeId })
    }

    override fun updateEmployee(e: Employee) {
        deleteEmployeeById(e.id)
        staticEmployeeList.add(e)
        staticEmployeeList.sort()
    }

    override fun getEmployeeById(employeeId: Int): Employee {
        return employeeList
                .stream()
                .filter { e -> e.id == employeeId }
                .findFirst()
                .get()
    }

    companion object {

        private var staticEmployeeList: MutableList<Employee> = ArrayList()

        private val sequence: IdSequence

        init {
            staticEmployeeList.add(Employee(1, "Michael", "Smith", "Mikey"))
            staticEmployeeList.add(Employee(2, "Jonathan", "Taylor", "Johnny"))
            staticEmployeeList.add(Employee(3, "David", "Wilson", "Dave"))
            sequence = IdSequence(4)
        }
    }

}
